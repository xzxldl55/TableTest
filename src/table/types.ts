/* eslint-disable @typescript-eslint/no-explicit-any */

import type { PropOptions, PropType } from 'vue-types/dist/types'
type Prop<T, D = T> = PropOptions<T, D> | PropType<T>
type PublicRequiredKeys<T> = {
  [K in keyof T]: T[K] extends { required: true } ? K : never
}[keyof T]

type PublicOptionalKeys<T> = Exclude<keyof T, PublicRequiredKeys<T>>
type InferPropType<T> = T extends null
  ? any // null & true would fail to infer
  : T extends { type: null | true }
    ? any // As TS issue https://github.com/Microsoft/TypeScript/issues/14829 // somehow `ObjectConstructor` when inferred from { (): T } becomes `any` // `BooleanConstructor` when inferred from PropConstructor(with PropMethod) becomes `Boolean`
    : T extends ObjectConstructor | { type: ObjectConstructor }
      ? Record<string, any>
      : T extends BooleanConstructor | { type: BooleanConstructor }
        ? boolean
        : T extends Prop<infer V, infer D>
          ? unknown extends V
            ? D
            : V
          : T

// eslint-disable-next-line @typescript-eslint/ban-types
export type IxPublicPropTypes<O> = O extends object
  ? { [K in PublicRequiredKeys<O>]: InferPropType<O[K]> } & { [K in PublicOptionalKeys<O>]?: InferPropType<O[K]> }
  : { [K in string]: any }

export interface paginationType {
  pageIndex: number
  pageSize: number
}

export type tableState = {
  data: Record<string, any>[]
  filters: string[]
  columns: ColumnPublicProps[]
  showHeader: boolean
}

export type TSortState = {
  name: string
  direction: 0 | 1
}

export type TFilterState = {
  filter: string
  searchValue: string
}

export interface TPageConf extends paginationType {
  count: number // 数据条目总数
}

// Props 定义在这里
export const tableProps = {
  data: { // 表格数据
    type: Array as PropType<Record<string, any>>,
    default: (): Array<Record<string, any>> => {
      return []
    },
    required: true,
  },
  columns: { // 列数据
    type: Array as PropType<ColumnPublicProps[]>,
    default: (): ColumnPublicProps[] => {
      return []
    },
    required: true,
  },
  showHeader: { // 是否显示表头
    type: Boolean,
    default: true,
  },
  pagination: { // 分页配置
    type: [Object as PropType<paginationType>, false],
    default: (): paginationType | false => {
      return false
    },
  },
  maxHeight: { // 表格体最大高度配置（默认按照内容撑开即高度为auto，配置后超出限制则滚动页面）
    type: [Number, String],
    default: -1,
  },
  filters: { // 可筛选字段
    type: Array,
    default: (): Array<string> => {
      return []
    },
  },
}
export const columnProps = {
  name: { // 列名 --> 即唯一标识id
    type: String,
    default: '',
    required: true,
  },
  title: { // 列标题
    type: String,
    default: '',
  },
  canSort: {
    type: Boolean,
    default: false,
  },
  renderFn: {
    type: Function || undefined,
    default: undefined,
  },
  sortState: {
    type: Object as PropType<TSortState>,
    default: (): TSortState => {
      return { name: '', direction: 0 }
    },
  },
}


export type TablePublicProps = IxPublicPropTypes<typeof tableProps>
export type ColumnPublicProps = IxPublicPropTypes<typeof columnProps>



