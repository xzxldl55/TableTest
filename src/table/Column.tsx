import { defineComponent, computed } from '@vue/composition-api';
import { columnProps, ColumnPublicProps } from './types';

export default defineComponent({
    name: 'TableColumn',
    props: columnProps,
    setup (props, { slots, emit }) {
        const title = props.title
        const name = props.name
        const canSort = props.canSort
        const classes = useClass(props)
        
        const emitSort = () => {
            emit('click', name);
        }
        return () => {
            return (
                <th column-name={name}>
                    { title }
                    {
                        canSort ?
                        // 区分一下上下高亮
                        <span class={classes.value}
                              onClick={emitSort}></span> :
                        ''
                    }
                </th>
            )
        }
    }
})

const useClass = (props: ColumnPublicProps) => {
    return computed(() => {
        return {
            'column-sort': true,
            'des': props.sortConf?.sortKey === props.name && !props.sortConf?.sort,
            'asc': props.sortConf?.sortKey === props.name && props.sortConf?.sort,
        }
    })
}