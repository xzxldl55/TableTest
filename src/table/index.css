.left {
    text-align: left;
}
.right {
    text-align: right;
}
.center {
    text-align: center;
}
.mr-1 {
    margin-right: 1rem;
}

/* 表格体 */
table {
    table-layout: fixed;
    border: 1px solid #d7dde4;
    font-size: 1rem;
    width: 100%;
    border-radius: 4px;
    box-sizing: border-box;
    color: #657180;
}
table > thead {
    background: #f5f7f9;
    color: #657180;
}
table > tbody {
    display: inline-block;
    width: 100%;
    overflow: auto;
}
tr {
    display: flex;
    width: 100%;
}
th, td {
    display: inline-block;
    flex: 1;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    padding: 8px 12px;
    text-align: center;
}

/* 列头 */
.column-sort {
    position: relative;
    display: inline-block;
    margin-left: 5px;
    cursor: pointer;
}
.column-sort::before { /* 升序图标 */
    position: absolute;
    top: -5px;
    content: '';
    border: 5px solid transparent;
    border-top-color: #657180;
}
.column-sort::after { /* 降序图标 */
    position: absolute;
    top: -18px;
    content: '';
    border: 5px solid transparent;
    border-bottom-color: #657180;
}
.column-sort.asc::before {
    border-top-color: #39f;
}
.column-sort.des::after {
    border-bottom-color: #39f;
}


/* 分页器 */
.pagination {
    padding: 8px 16px;
    border-top: 1px solid #e3e8ee;
}
.pagination-btn {
    display: inline-block;
    width: 32px;
    height: 32px;
    text-align: center;
    line-height: 32px;
    color: #657180;
    border: 1px solid #d7dde4;
    margin-right: 8px;
    border-radius: 4px;
    cursor: pointer;
    transition: .5s;
}
.pagination-btn:last-child {
    margin-right: 0;
}
.pagination-btn:hover {
    border-color: #39f;
    color: #39f;
}
.pagination-btn.on {
    background: #39f;
    color: #fff;
    border-color: transparent;
}