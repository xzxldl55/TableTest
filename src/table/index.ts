/**
 * Created by uedc on 2021/10/11.
 */

import TestTable from './Table'
import TestColumn from './Column'
import './index.css'

export {
  TestTable,
  TestColumn,
}
