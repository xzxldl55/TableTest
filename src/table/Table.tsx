/**
 * Created by uedc on 2021/10/11.
 */

import { computed, defineComponent, reactive, ref, watch } from '@vue/composition-api'
import { TablePublicProps, tableProps, paginationType, ColumnPublicProps } from './types'
import Column from './Column'
import { sortBy } from 'lodash-es'
import debug from 'debug'

const DEBUG = debug('table:body')

type tableState = {
  sortConf: {
    sortKey: string
    sort: boolean 
  }
  data: Record<string, any>[]
  filter: string
  searchValue: string
  filters: string[]
  pagination: false | { pageIndex: number; pageSize: number }
  columns: ColumnPublicProps[]
  count: number
}

export default defineComponent({
  name: 'Table',
  props: tableProps,
  setup(props: TablePublicProps, { slots, emit }) {
    const style = useStyle(props)
    const state: tableState = reactive({
      sortConf: { // 当前表格排序配置（升序/降序）
        sortKey: '',
        sort: false // true升序，false降序
      },
      data: props.data as Record<string, any>[],
      filter: props.filters?.length ? (props.filters[0] as string) : '',
      searchValue: '',
      filters: props.filters as unknown as string[],
      pagination: props.pagination as unknown as paginationType | false,
      columns: props.columns as ColumnPublicProps[],
      count: 0, // 当前数据条目总数（过滤后）
    })
    const tableData = useTableData(state) // 获取当前页table数据【关联props.data, 分页，排序，筛选】
    const size = computed(() => { // 定义当前分页总数
      if (state.pagination) {
        return Math.ceil(state.count / state.pagination.pageSize)
      }
      return 0
    })

    // 监听数据变动，输出日志
    watch(tableData, (): void => {
      DEBUG('data change')
    })


    // 总页数

    // 【排序】
    const sortTable = (name: string) => {

      // 回到首页
      if (state.pagination) {
        state.pagination.pageIndex = 1
      }

      // 当前sortBy非name，则初始化按照name升序排序
      if (state.sortConf.sortKey !== name) {
        state.sortConf.sortKey = name
        state.sortConf.sort = true
      } else { // 正按照当前列排序，将排序方式取反
        state.sortConf.sort = !state.sortConf.sort
      }

      DEBUG('sort by', name, state.sortConf.sort ? '升序' : '降序')

      // 抛出排序事件
      emit('sort', { name, sort: state.sortConf.sort })
    }

    // 【切换页码】
    const changePage = (index: number) => {
      if (!state.pagination) {
        return
      }
      if (state.pagination.pageIndex === index) {
        return
      }
      state.pagination.pageIndex = index;

      DEBUG('page change', index)

      // 页码切换事件
      emit('page-change', index)
    }

    return () => {
      const operators = slots.operator
      return (
        <div class="simple-table">
          {
            renderHeader(state.filters, state)
          }
          <table>
            {/* 表头 */}
            <thead vShow={props.showHeader}>
              <tr>
                {
                  state.columns.map(column => {
                    
                    // 可支持slot[columns.name]自定义配置表头
                    return slots[column.name] ?
                      <th>{ slots[column.name]({ column }) }</th> :
                      <Column title={column.title}
                              name={column.name}
                              canSort={column.canSort}
                              sortConf={state.sortConf}
                              onClick={sortTable} />
                  })
                }
                {
                  operators ?
                  <th>操作</th> :
                  ''
                }
              </tr>
            </thead>
            {/* 表格主体 */}
            <tbody style={style.value}>
              {
                tableData.value.map(item => {
                  return (<tr>
                      {
                        state.columns.map(column => {
                          const content = (column.renderFn ? column.renderFn(item[column.name]) : item[column.name]) || '-'
                          return <td column-data-name={column.name}
                                    title={content}>{ content }</td>
                        })
                      }
                      <td class="center">
                        { operators ? operators({ item }) : '' }
                      </td>
                    </tr>)
                })
              }
            </tbody>
            {/* 分页器 */}
            <div class="pagination center">
              {
                state.pagination ? 
                (
                  new Array(size.value).fill(1).map((v, i) => {
                    return <button class={['pagination-btn', { 'on': state.pagination.pageIndex === i + 1 }]}
                                  onClick={changePage.bind(null, i + 1)}>{ i + 1 }</button>
                  })
                ) :
                ''
              }
            </div>
          </table>
        </div>
      )
    }
  },
})

// tBody样式处理
function useStyle (props: TablePublicProps) {
  return computed(() => {
    return {
      maxHeight: typeof props.maxHeight === 'string' ? props.maxHeight : props.maxHeight + 'px'
    }
  })
}

// 返回应渲染的Table数据（在函数内进行对数据处理，如分页、排序，筛选等）
function useTableData (state: tableState) {
  return computed(() => {
    DEBUG('data change')

    let resData = state.data || []
    const {
      pagination,
      sortConf,
      filter,
      searchValue
    } = state

    // 1. filter & searchValue 按照关键词过滤数据
    if (filter && searchValue) {
      const renderFn = state.columns.filter(column => column.name === filter)[0].renderFn;
      resData = resData.filter(data => {

        // 支持renderFn的过滤
        const renderData = (renderFn ? renderFn(data[filter]) : data[filter]).toString()
        return renderData.toLocaleLowerCase()
          .includes(searchValue.toLocaleLowerCase())
      })
    }

    // 2. 更新数据条目总数（PS：注意需要放在最后一个对resData赋值语句之后 return 跳出之前，确保数据准确性）
    state.count = resData.length

    // 3. 不分页 ｜ 空数据直接返回
    if (!resData || !pagination) {
      return resData
    }

    // ... TODO：ect 其他功能添加，如 已选列表prop的处理

    // 4. 获取当前分页配置
    const start = (pagination.pageIndex - 1) * pagination.pageSize
    const end = start + pagination.pageSize

    // 5. 检测是否需要按某列排序
    if (sortConf.sortKey) {
      return sortConf.sort ?
        sortBy(resData, (obj: Record<string, any>) => obj[sortConf.sortKey]).slice(start, end) :
        sortBy(resData, (obj: Record<string, any>) => obj[sortConf.sortKey]).reverse().slice(start, end)
    }
    return resData.slice(start, end)
  })
}

// 表格工具栏渲染
function renderHeader (filters: string[], state: tableState) {
  return (
    <header>
      {
        filters.length ?
        <div class="table-search-container right">
          {/* TODO：无法使用v-model指令？待解决 */}
          <select class="table-search-select"
                  name="tableFilter"
                  value={state.filter}
                  onChange={(e: Event) => {
                    const target = e.target as HTMLSelectElement
                    state.filter = target.value

                    // 切换过滤器后自动清空搜索关键词
                    state.searchValue = ''
                  }}>
            {
              filters.map(filter => <option value={filter}>{filter}</option>)
            }
          </select>
          {/* TODO：无法使用v-model指令？待解决 */}
          <input class="table-search-input"
                 type="text"
                 value={state.searchValue}
                 onKeydown={search.bind(null, state)} />
        </div>
        : ''
      }
    </header>
  )
}

const search = (
  state: tableState,
  e: KeyboardEvent
) => {
  const target = e.target as HTMLInputElement
  if (e.code === 'Enter') {
    state.searchValue = target.value
    DEBUG('filter table data', state.filter, state.searchValue)
  }
}