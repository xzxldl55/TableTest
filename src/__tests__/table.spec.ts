import { mount } from '@vue/test-utils'
import { TestTable } from '../table/index'
import Vue from 'vue'
import CompositionApi from '@vue/composition-api'
import {
  Staff,
  mockTableColumns,
  mockTableData,
} from '../mockData'
import { ColumnPublicProps, TablePublicProps, paginationType } from 'src/table/types'

Vue.use(CompositionApi)


const staffMockData: Staff[] = mockTableData
const tableColumns: ColumnPublicProps = mockTableColumns
const tablePagination = {
  pageSize: 10, // 每页大小
  pageIndex: 1, // 当前页码
}
const defaultProps: TablePublicProps = {
  data: staffMockData.slice(0, 10),
  columns: tableColumns as ColumnPublicProps,
  pagination: tablePagination as paginationType,
  maxHeight: 240,
  showHeader: true,
  filters: ['name', 'age', 'sex'],
}
describe('Table', () => {
  const TableMount = (options: Record<string, unknown>) => mount(TestTable, options)
  test('开始测试', () => {
    expect(1).toBe(1)
  })

  test('render Default', () => {
    const wrapper = TableMount({ propsData: defaultProps })
    expect(wrapper.html()).toMatchSnapshot()
    expect(() => {
      wrapper.vm.$forceUpdate()
      wrapper.vm.$destroy()
    }).not.toThrow()
  })

  test('props test', async () => {
    const wrapper = TableMount({
      propsData: {
        test: true,
        ...defaultProps,
      },
    })

    // 1. 分页配置测试 --> 校验pageIndex/pageSize设置结果
    expect(wrapper.findAll('.simple-table tbody > tr').length).toEqual(10)
    expect(wrapper.find('.pagination-btn.on').element.innerHTML).toEqual('1')

    // 2. maxHeight最大高度测试，获取Tbody实际高度
    expect(wrapper.find('.simple-table tbody').attributes('style')).toContain('max-height: 240px')

    // 3. 列数据测试，传入renderFn，校验实际结果
    expect(wrapper.findAll('.simple-table td[column-data-name=sex]').filter(node => !(['男', '女'].includes(node.element.innerHTML))).length).toEqual(0)
  })

  test('sort data', async () => {
    const wrapper = TableMount({
      propsData: defaultProps,
    })

    {

      // 按年龄排序
      await wrapper.find('.simple-table th[column-name=age] .column-sort').trigger('click') // 模拟点击排序
      const ageData = wrapper.findAll('.simple-table td[column-data-name=age]').wrappers.map(node => Number(node.element.innerHTML))
      const ageSortBeforeStr = ageData.toString()

      // 检查是否升序排序了
      expect(ageSortBeforeStr).toEqual(ageData.sort((a, b) => a - b).toString())
    }

    {
      // 按年龄排序
      await wrapper.find('.simple-table th[column-name=age] .column-sort').trigger('click') // 模拟点击排序
      const ageData = wrapper.findAll('.simple-table td[column-data-name=age]').wrappers.map(node => Number(node.element.innerHTML))
      const ageSortBeforeStr = ageData.toString() // 将结果转换成字符串方便比较

      // 检查是否降序排序了
      expect(ageSortBeforeStr).toEqual(ageData.sort((a, b) => b - a).toString())
    }
  })

  test('filter data test', async () => {
    const wrapper = TableMount({
      propsData: defaultProps,
    })

    // 输入关键词45按name进行搜索
    const inputEle = wrapper.find('.table-search-input')
    inputEle.element.setAttribute('value', '45')
    await inputEle.trigger('keydown.enter') // TODO：触发无效？

    // expect(Array.from(wrapper.findAll('.simple-table td[column-data-name=name]').wrappers).every(v => v.element.innerHTML.toLocaleLowerCase().includes('45'))).toBeTruthy()

  })

  test('pagination test', async () => {
    const startTime = Date.now()
    const wrapper = TableMount({
      propsData: {
        ...defaultProps,
        data: mockTableData, // 500数据总量
        pagination: {
          pageIndex: 2,
          pageSize: 100,
        },
      },
    })
    const endTime = Date.now()

    // 测试100条数据的渲染速度 < 500ms
    expect(endTime - startTime < 500).toBeTruthy()

    // 检测分页配置是否有问题
    expect(wrapper.find('.simple-table .pagination-btn.on').element.innerHTML).toEqual('2')
    expect(wrapper.findAll('.simple-table tbody > tr').length).toEqual(100)

    // 切换页码 -- 跳转到第五页
    await wrapper.find('.simple-table .pagination-btn:nth-child(6)').trigger('click')
    expect(wrapper.find('.simple-table .pagination-btn.on').element.innerHTML).toEqual('5')
  })
})
