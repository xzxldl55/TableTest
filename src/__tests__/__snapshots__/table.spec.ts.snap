// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table render Default 1`] = `
"<div class=\\"simple-table\\">
  <header>
    <div class=\\"table-search-container right\\"><select name=\\"tableFilter\\" class=\\"table-search-select\\">
        <option value=\\"name\\">name</option>
        <option value=\\"age\\">age</option>
        <option value=\\"sex\\">sex</option>
      </select><input type=\\"text\\" class=\\"table-search-input\\"></div>
  </header>
  <table>
    <thead>
      <tr>
        <th column-name=\\"name\\">姓名<span class=\\"column-sort\\"></span></th>
        <th column-name=\\"age\\">年龄<span class=\\"column-sort\\"></span></th>
        <th column-name=\\"sex\\">性别</th>
      </tr>
    </thead>
    <tbody style=\\"max-height: 240px;\\">
      <tr>
        <td column-data-name=\\"name\\" title=\\"Jerry0\\">Jerry0</td>
        <td column-data-name=\\"age\\" title=\\"1\\">1</td>
        <td column-data-name=\\"sex\\" title=\\"女\\">女</td>
        <td class=\\"center\\"></td>
      </tr>
      <tr>
        <td column-data-name=\\"name\\" title=\\"Jerry1\\">Jerry1</td>
        <td column-data-name=\\"age\\" title=\\"2\\">2</td>
        <td column-data-name=\\"sex\\" title=\\"男\\">男</td>
        <td class=\\"center\\"></td>
      </tr>
      <tr>
        <td column-data-name=\\"name\\" title=\\"Jerry2\\">Jerry2</td>
        <td column-data-name=\\"age\\" title=\\"3\\">3</td>
        <td column-data-name=\\"sex\\" title=\\"女\\">女</td>
        <td class=\\"center\\"></td>
      </tr>
      <tr>
        <td column-data-name=\\"name\\" title=\\"Jerry3\\">Jerry3</td>
        <td column-data-name=\\"age\\" title=\\"4\\">4</td>
        <td column-data-name=\\"sex\\" title=\\"男\\">男</td>
        <td class=\\"center\\"></td>
      </tr>
      <tr>
        <td column-data-name=\\"name\\" title=\\"Jerry4\\">Jerry4</td>
        <td column-data-name=\\"age\\" title=\\"5\\">5</td>
        <td column-data-name=\\"sex\\" title=\\"女\\">女</td>
        <td class=\\"center\\"></td>
      </tr>
      <tr>
        <td column-data-name=\\"name\\" title=\\"Jerry5\\">Jerry5</td>
        <td column-data-name=\\"age\\" title=\\"6\\">6</td>
        <td column-data-name=\\"sex\\" title=\\"男\\">男</td>
        <td class=\\"center\\"></td>
      </tr>
      <tr>
        <td column-data-name=\\"name\\" title=\\"Jerry6\\">Jerry6</td>
        <td column-data-name=\\"age\\" title=\\"7\\">7</td>
        <td column-data-name=\\"sex\\" title=\\"女\\">女</td>
        <td class=\\"center\\"></td>
      </tr>
      <tr>
        <td column-data-name=\\"name\\" title=\\"Jerry7\\">Jerry7</td>
        <td column-data-name=\\"age\\" title=\\"8\\">8</td>
        <td column-data-name=\\"sex\\" title=\\"男\\">男</td>
        <td class=\\"center\\"></td>
      </tr>
      <tr>
        <td column-data-name=\\"name\\" title=\\"Jerry8\\">Jerry8</td>
        <td column-data-name=\\"age\\" title=\\"9\\">9</td>
        <td column-data-name=\\"sex\\" title=\\"女\\">女</td>
        <td class=\\"center\\"></td>
      </tr>
      <tr>
        <td column-data-name=\\"name\\" title=\\"Jerry9\\">Jerry9</td>
        <td column-data-name=\\"age\\" title=\\"10\\">10</td>
        <td column-data-name=\\"sex\\" title=\\"男\\">男</td>
        <td class=\\"center\\"></td>
      </tr>
    </tbody>
    <div class=\\"pagination center\\"><button class=\\"pagination-btn on\\">1</button></div>
  </table>
</div>"
`;
